File: src/api/dashboard.service.ts
import api from './axios';
import { DashboardSummary, LowStockProduct, MovementChartItem } from '@/types';

export const dashboardService = {
  async getSummary(): Promise<DashboardSummary> {
    try {
      const { data } = await api.get('/dashboard/summary');
      return {
        totalProducts: data.total_products ?? data.totalProducts ?? 0,
        totalItemsInStock: data.total_items_in_stock ?? data.totalItemsInStock ?? 0,
        lowStockProductsCount: data.total_low_stock_products ?? data.low_stock_products_count ?? data.lowStockProductsCount ?? 0,
        totalEntries: data.total_entries ?? data.totalEntries ?? 0,
        totalExits: data.total_exits ?? data.totalExits ?? 0,
      };
    } catch {
      // Fallback: compute from products and movements
      try {
        const [productsRes, movementsRes] = await Promise.all([
          api.get('/products', { params: { size: 100 } }),
          api.get('/stock-movements', { params: { size: 100 } }),
        ]);

        const prods = Array.isArray(productsRes.data) ? productsRes.data : (productsRes.data?.items || []);
        const movs = Array.isArray(movementsRes.data) ? movementsRes.data : (movementsRes.data?.items || []);

        const totalItems = prods.reduce((s: number, p: any) => s + (p.quantity ?? 0), 0);
        const lowStock = prods.filter((p: any) => p.quantity <= (p.minimum_stock ?? p.minimumStock ?? 0));
        const entries = movs
          .filter((m: any) => ['IN', 'ENTRY'].includes(m.movement_type || m.movementType))
          .reduce((s: number, m: any) => s + (m.quantity ?? 0), 0);
        const exits = movs
          .filter((m: any) => ['OUT', 'EXIT'].includes(m.movement_type || m.movementType))
          .reduce((s: number, m: any) => s + (m.quantity ?? 0), 0);

        return {
          totalProducts: prods.length,
          totalItemsInStock: totalItems,
          lowStockProductsCount: lowStock.length,
          totalEntries: entries,
          totalExits: exits,
        };
      } catch {
        return {
          totalProducts: 0,
          totalItemsInStock: 0,
          lowStockProductsCount: 0,
          totalEntries: 0,
          totalExits: 0,
        };
      }
    }
  },

  async getLowStockProducts(): Promise<LowStockProduct[]> {
    try {
      // Correct endpoint: /dashboard/low-stock
      const { data } = await api.get('/dashboard/low-stock');
      const items = data?.items || data || [];
      return (Array.isArray(items) ? items : []).map((p: any) => ({
        id: String(p.id),
        name: p.name,
        sku: p.sku || '',
        quantity: p.quantity ?? 0,
        minimumStock: p.minimum_stock ?? p.minimumStock ?? 0,
      }));
    } catch {
      // Fallback: filter from all products
      try {
        const { data } = await api.get('/products', { params: { size: 100 } });
        const prods = Array.isArray(data) ? data : (data?.items || []);
        return prods
          .filter((p: any) => p.quantity <= (p.minimum_stock ?? p.minimumStock ?? 0))
          .map((p: any) => ({
            id: String(p.id),
            name: p.name,
            sku: p.sku || '',
            quantity: p.quantity ?? 0,
            minimumStock: p.minimum_stock ?? p.minimumStock ?? 0,
          }));
      } catch {
        return [];
      }
    }
  },

  async getMovementChart(): Promise<MovementChartItem[]> {
    try {
      // Correct endpoint: /dashboard/movements
      const { data } = await api.get('/dashboard/movements');
      const movements = data?.movements || data || [];
      if (Array.isArray(movements) && movements.length > 0) {
        return movements.map((m: any) => ({
          date: m.date,
          entries: m.entries ?? 0,
          exits: m.exits ?? 0,
        }));
      }
      throw new Error('No data');
    } catch {
      // Fallback: compute from movements
      try {
        const { data } = await api.get('/stock-movements', { params: { size: 100 } });
        const movs = Array.isArray(data) ? data : (data?.items || []);
        const days = ['Dom', 'Seg', 'Ter', 'Qua', 'Qui', 'Sex', 'Sáb'];
        const chartData = days.map((d) => ({ date: d, entries: 0, exits: 0 }));
        movs.forEach((m: any) => {
          const day = new Date(m.created_at || m.createdAt).getDay();
          const type = m.movement_type || m.movementType;
          if (['IN', 'ENTRY'].includes(type)) chartData[day].entries += m.quantity ?? 0;
          if (['OUT', 'EXIT'].includes(type)) chartData[day].exits += m.quantity ?? 0;
        });
        return chartData;
      } catch {
        return [];
      }
    }
  },
};

File: backend/app/services/dashboard_service.py
"""
Dashboard Service — aggregated analytics for the inventory system.
"""

from typing import Optional

import structlog
from sqlalchemy.orm import Session

from app.repositories.product_repository import ProductRepository
from app.repositories.stock_movement_repository import StockMovementRepository
from app.schemas.dashboard import (
    DashboardMovementsResponse,
    DashboardSummary,
    LowStockProduct,
    LowStockResponse,
    MovementSummary,
)
from app.utils.enums import MovementType

logger = structlog.get_logger(__name__)


class DashboardService:
    """Service layer for dashboard analytics."""

    def __init__(self, db: Session) -> None:
        self._product_repo = ProductRepository(db)
        self._movement_repo = StockMovementRepository(db)

    def get_summary(self) -> DashboardSummary:
        """Build the aggregated dashboard summary."""
        return DashboardSummary(
            total_products=self._product_repo.count_all(),
            total_items_in_stock=self._product_repo.total_items_in_stock(),
            total_low_stock_products=self._product_repo.count_low_stock(),
            total_entries=self._movement_repo.count_by_type(MovementType.ENTRY),
            total_exits=self._movement_repo.count_by_type(MovementType.EXIT),
            total_adjustments=self._movement_repo.count_by_type(MovementType.ADJUSTMENT),
            total_stock_value=self._product_repo.total_stock_value(),
        )

    def get_movements_by_period(self, days: int = 30) -> DashboardMovementsResponse:
        """Get movement aggregation by date for the given period."""
        raw = self._movement_repo.get_movements_by_period(days=days)

        movements = [
            MovementSummary(
                date=item["date"],
                entries=item["entries"],
                exits=item["exits"],
                adjustments=item["adjustments"],
            )
            for item in raw
        ]

        period_start = movements[0].date if movements else None
        period_end = movements[-1].date if movements else None

        return DashboardMovementsResponse(
            movements=movements,
            period_start=period_start,
            period_end=period_end,
        )

    def get_low_stock_products(self) -> LowStockResponse:
        """Get all products with stock at or below minimum threshold."""
        products = self._product_repo.get_low_stock()

        items = [
            LowStockProduct(
                id=p.id,
                name=p.name,
                sku=p.sku,
                category=p.category,
                quantity=p.quantity,
                minimum_stock=p.minimum_stock,
                deficit=p.minimum_stock - p.quantity,
            )
            for p in products
        ]

        return LowStockResponse(items=items, total=len(items))
